### PR TITLE
Add open source licenses screen to Android app

### LINKED ISSUE #1471: Support showing licenses in About page
We can use [Licensee](https://github.com/cashapp/licensee) to generate a dependency report and show it in About page.

### PR DESCRIPTION
**Pull request by penkzhou**
current only enable on QaDebug flavor on android platform.

Test path:

1. execute ':android-app:app:generateLicensesAsset' task
2. run the project with android device.
3. go to the settings screen.
4. click the "View Open Source Licenses" item
5. check the "Open Source Licenses" page.

screenshot(video):

[Watch the video](https://github.com/chrisbanes/tivi/assets/1454424/b42678eb-1279-4882-9e40-305a39fcb49c)

Fixes #1471

### CODE REVIEW

File: android-app/app/src/main/assets/generated_licenses.json
Comment (chrisbanes) [inline android-app/app/src/main/assets/generated_licenses.json:1]: We should not commit this file

File: data/opensource/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    id("app.tivi.android.library")
+    id("app.tivi.kotlin.multiplatform")
+    kotlin("plugin.serialization") version "1.9.10"
+}
+
+kotlin {
+    sourceSets {
+        val commonMain by getting {
+            dependencies {
+                api(projects.core.base)
```
Comment (chrisbanes) [inline data/opensource/build.gradle.kts:18]: This module doesn't need all of these dependencies.

File: data/opensource/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+                implementation(libs.androidx.collection) // LruCache
+
+                implementation(libs.kotlininject.runtime)
+            }
+        }
+
+        val androidMain by getting {
+            dependencies {
+                api(libs.appauth)
```
Comment (chrisbanes) [inline data/opensource/build.gradle.kts:39]: Ditto, can remove all? of these.

File: data/opensource/src/androidMain/kotlin/app/tivi/data/opensource/store/AndroidOpenSourceStore.kt [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.asFlow
+import kotlinx.coroutines.flow.emptyFlow
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.json.Json
+import me.tatarka.inject.annotations.Inject
+
+@Inject
+class AndroidOpenSourceStore(private val context: Application) : OpenSourceStore {
+    private var lastAuthState: OpenSourceState = OpenSourceState.Empty
```
Comment (chrisbanes) [inline data/opensource/src/androidMain/kotlin/app/tivi/data/opensource/store/AndroidOpenSourceStore.kt:21]: Rename to `lastLicensesState`

File: ui/opensource/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+android {
+    namespace = "app.tivi.opensource"
+}
+
+kotlin {
+    sourceSets {
+        val commonMain by getting {
+            dependencies {
+                implementation(projects.core.base)
+                implementation(projects.core.preferences)
```
Comment (chrisbanes) [inline ui/opensource/build.gradle.kts:20]: Don't need preferences

File: .idea/dictionaries/penkzhou.xml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+<component name="ProjectDictionaryState">
```
Comment (chrisbanes) [inline .idea/dictionaries/penkzhou.xml:1]: Please delete

File: .idea/migrations.xml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (chrisbanes) [inline .idea/migrations.xml:1]: Please delete

File: android-app/app/build.gradle.kts [hunk 7]
```diff
@@ -178,6 +186,45 @@ if (file("google-services.json").exists()) {
     }
 }
 
+abstract class GenerateLicensesAsset : DefaultTask() {
```
Comment (chrisbanes) [inline android-app/app/build.gradle.kts:189]: Please move this to `gradle/build-logic`

File: data/opensource/src/androidMain/kotlin/app/tivi/data/opensource/store/AndroidOpenSourceStore.kt [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+    @OptIn(ExperimentalSerializationApi::class)
+    fun getOpenSourceItemList(): Flow<List<OpenSourceItem>> {
+        try {
+            val inputStream: InputStream = context.assets.open("generated_licenses.json")
+            val json = Json {
+                ignoreUnknownKeys = true
+                explicitNulls = false
+            }
+            return inputStream.bufferedReader().lineSequence().map { json.decodeFromString<List<OpenSourceItem>>(it).toMutableList() }.asFlow()
```
Comment (chrisbanes) [inline data/opensource/src/androidMain/kotlin/app/tivi/data/opensource/store/AndroidOpenSourceStore.kt:35]: This can be replaced with [Json](https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-json/kotlinx.serialization.json/-json/index.html).[decodeFromStream](https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-json/kotlinx.serialization.json/decode-from-stream.html):

```
Json.decodeFromStream<List<OpenSourceItem>>(inputStream)
```

File: data/opensource/src/androidMain/kotlin/app/tivi/data/opensource/store/AndroidOpenSourceStore.kt [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+@Inject
+class AndroidOpenSourceStore(private val context: Application) : OpenSourceStore {
+    private var lastAuthState: OpenSourceState = OpenSourceState.Empty
+
+    override fun fetch(): OpenSourceState {
+        return SimpleOpenSourceState(getOpenSourceItemList())
+    }
+
+    @OptIn(ExperimentalSerializationApi::class)
+    fun getOpenSourceItemList(): Flow<List<OpenSourceItem>> {
```
Comment (chrisbanes) [inline data/opensource/src/androidMain/kotlin/app/tivi/data/opensource/store/AndroidOpenSourceStore.kt:28]: There's no need for this to be a Flow. This should be a suspending function with uses `withContext(dispatchers.io) { ... } `

File: data/opensource/src/commonMain/kotlin/app/tivi/data/opensource/OpenSourceState.kt [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+package app.tivi.data.opensource
+
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.emptyFlow
+import kotlinx.serialization.Serializable
+
+interface OpenSourceState {
+    val openSourceList: Flow<List<OpenSourceItem>>
+    fun serializeToJson(): String
```
Comment (chrisbanes) [inline data/opensource/src/commonMain/kotlin/app/tivi/data/opensource/OpenSourceState.kt:12]: What is this used for?

File: data/opensource/src/commonMain/kotlin/app/tivi/data/opensource/OpenSourceState.kt [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+package app.tivi.data.opensource
+
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.emptyFlow
+import kotlinx.serialization.Serializable
+
+interface OpenSourceState {
+    val openSourceList: Flow<List<OpenSourceItem>>
```
Comment (chrisbanes) [inline data/opensource/src/commonMain/kotlin/app/tivi/data/opensource/OpenSourceState.kt:11]: As above, this should be a `val items: List<OpenSourceItem>`

File: data/opensource/src/commonMain/kotlin/app/tivi/data/opensource/OpenSourceState.kt [hunk 12]
```diff
@@ -0,0 +1,66 @@
+// Copyright 2023, Google LLC, Christopher Banes and the Tivi project contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package app.tivi.data.opensource
+
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.emptyFlow
+import kotlinx.serialization.Serializable
+
+interface OpenSourceState {
```
Comment (chrisbanes) [inline data/opensource/src/commonMain/kotlin/app/tivi/data/opensource/OpenSourceState.kt:10]: State should be a `data class`, not an interface

File: data/opensource/src/androidMain/kotlin/app/tivi/data/opensource/store/AndroidOpenSourceStore.kt [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+import java.io.InputStream
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.asFlow
+import kotlinx.coroutines.flow.emptyFlow
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.json.Json
+import me.tatarka.inject.annotations.Inject
+
+@Inject
+class AndroidOpenSourceStore(private val context: Application) : OpenSourceStore {
```
Comment (chrisbanes) [inline data/opensource/src/androidMain/kotlin/app/tivi/data/opensource/store/AndroidOpenSourceStore.kt:20]: Only the `getOpenSourceItemList()` function here will actually be different on each platform. Please move this to `commonMain` and then extract a `LicensesFetcher` interface which is responsible for loading the JSON file and returning a `List<OpenSourceItem>`. Each platform will need it's own `LicensesFetcher` impl.

File: settings.gradle.kts [hunk 14]
```diff
@@ -6,16 +6,16 @@ pluginManagement {
     includeBuild("gradle/build-logic")
 
     repositories {
-        gradlePluginPortal()
```
Comment (chrisbanes) [inline settings.gradle.kts:9]: Why this change?

File: settings.gradle.kts [hunk 15]
```diff
@@ -6,16 +6,16 @@ pluginManagement {
...
-        google()
         mavenCentral()
+        google()
+        gradlePluginPortal()
     }
 }
 
 dependencyResolutionManagement {
     repositories {
-        google()
```
Comment (chrisbanes) [inline settings.gradle.kts:17]: Ditto, why this change?

File: ui/opensource/src/commonMain/kotlin/app/tivi/settings/opensource/OpenSource.kt [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+            )
+        },
+        contentWindowInsets = WindowInsets.systemBars,
+        modifier = modifier,
+    ) { contentPadding ->
+        LazyColumn(
+            contentPadding = contentPadding,
+            modifier = Modifier.fillMaxWidth(),
+        ) {
+            state.opensourceItemList.forEach { item ->
```
Comment (chrisbanes) [inline ui/opensource/src/commonMain/kotlin/app/tivi/settings/opensource/OpenSource.kt:77]: Use `items(state.opensourceItemList)`

**Review by chrisbanes — COMMENTED** [review]
Great start, thanks! Added some comments below.

Also, I think we should rename the packages and modules to `licenses` rather than `opensource`. It's a lot clearer imo.

File: android-app/app/build.gradle.kts [hunk 17]
```diff
@@ -178,6 +189,71 @@ if (file("google-services.json").exists()) {
...
+        val owner = parts.getOrNull(0) ?: return null
+        val name = parts.getOrNull(1) ?: return null
+        return owner to name
+    }
+}
+
+val generateLicenseTask =
+    tasks.register<GenerateLicensesAsset>("generateLicensesAsset") {
+        buildDir.set(project.layout.buildDirectory)
+        jsonFile.set(project.layout.projectDirectory.file("src/main/assets/generated_licenses.json"))
```
[the remark this answers is not in the document]
Comment (penkzhou) [reply android-app/app/build.gradle.kts:216]: Ok. I will take a look at it later.

File: settings.gradle.kts (hunk 14 above)
Comment (penkzhou) [reply settings.gradle.kts:9]: the mavenCentral() host the most library so I place it first.

File: data/opensource/src/commonMain/kotlin/app/tivi/data/opensource/OpenSourceState.kt (hunk 10 above)
Comment (penkzhou) [reply data/opensource/src/commonMain/kotlin/app/tivi/data/opensource/OpenSourceState.kt]: sorry for that. This code is writing by following the code style in other module.
I have change it.

**Comment by chrisbanes** [comment]
Thanks for the changes! I think this is ready to go in. Can you please merge in `main` as that contains some fixes for CI runs from forks?
